src/discussion.cairo

```diff
@@ -0,0 +1,135 @@
+    pub struct Comment {
+        user: ContractAddress,
+        ipfs_hash: felt252,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {}
+
+    #[embeddable_as(DiscussionImpl)]
+    impl Discussions<TContractState, +HasComponent<TContractState>,
+    +Drop<TContractState>,
+    impl Proposals: proposals_component::HasComponent<TContractState>,
+    > of super::IDiscussion<ComponentState<TContractState>> {
+        fn add_comment(ref self: ComponentState<TContractState>, prop_id: felt252, ipfs_hash: felt252) {
+            //Check if proposal is live 
+            let is_live = self.is_proposal_live(prop_id);
+            assert(is_live == 1, 'Proposal is not live!');
+
+            //Check if caller is a CRAM token holder
+            let user_address = get_caller_address();
+            let govtoken_addr = get_governance_token_address_self();
+            let caller_balance : u256 = IERC20Dispatcher { contract_address: govtoken_addr }
+                .balanceOf(user_address);
+            assert(caller_balance.high != 0 || caller_balance.low != 0, 'Govtoken balance is zero');
+
+            //get current comment count 
+            let count: u64 = self.comment_count.read(prop_id);
+
+            //store new comment/ipfs_hash at next index
+            let new_comment = Comment {user: user_address, ipfs_hash: ipfs_hash};
+            self.comments.write((prop_id, count), new_comment);
+
+            //Increment comment count for proposal by one
+            self.comment_count.write(prop_id, count + 1);
+            
+        }
+
+        fn get_comment(self: @ComponentState<TContractState>, prop_id: felt252) -> Array<(ContractAddress, felt252)> {
+            //Get comment counts 
+            let count: u64 = self.comment_count.read(prop_id);
+
+            //Initialize an array of comments
+            let mut arr = ArrayTrait::<(ContractAddress, felt252)>::new();
+
+            //if no comments, return empty array
+            if count == 0 {
+                return arr;
+            }
+
+            // loop over comment count and collect comments
+            let mut i: u64 = 0;
+            loop {
+                if i >= count {
+                    break;
+                }
+
+                //collect comment at position i
+                let com: Comment = self.comments.read((prop_id, i));
+                arr.append((com.user, com.ipfs_hash));
+                i += 1;  
+            };
+
+            // return array of comments
+            arr
+        }
+    }
+
+    #[generate_trait]
+    impl InternalImpl<TContractState, +HasComponent<TContractState>,
+    +Drop<TContractState>,
+    impl Proposals: proposals_component::HasComponent<TContractState>,
+    > of InternalTrait<TContractState> {
+        fn is_proposal_live(ref self: ComponentState<TContractState>, prop_id: felt252 ) -> u8 {
+
+            let proposals_comp = get_dep_component!(@self, Proposals);
+
+            //Get live proposals
```

Comment:
Comments are to be used to clarify unclear parts of the code, this is superfluous